cpp/src/arrow/acero/options.h

```diff
@@ -52,13 +52,21 @@ class Executor;
 
 namespace acero {
 
+#ifndef NDEBUG
+struct DebugOptions;
+#endif
+
 using AsyncExecBatchGenerator = AsyncGenerator<std::optional<ExecBatch>>;
 
 /// \addtogroup execnode-options
 /// @{
 class ARROW_ACERO_EXPORT ExecNodeOptions {
  public:
   virtual ~ExecNodeOptions() = default;
+
+#ifndef NDEBUG
+  std::shared_ptr<DebugOptions> debug_opts;
+#endif
```

Comment:
This makes the ABI depend on `NDEBUG`, which should really be avoided. Some people do use non-debug versions of Arrow in debug builds of their code (and potentially vice-versa, though of course much less common).

Reply:
I'm not sure how to solve this nicely. Could you make a suggestion? Not-so-nice solutions I see are:
1. Change to a `std::shared_ptr<void*>`.
2. Expose `DebugOptions` in release-mode too (but not use it).

Reply:
Given I don't understand what `DebugOptions`, it's hard to answer. But I would question why `DebugOptions` would not be used in release mode. What is the use case?

Reply:
`DebugOptions` holds a debug-stream and a debug-mutex for it that are only used in debug-mode for collecting debug-messages specific to a node. With the [recent commit](https://github.com/apache/arrow/pull/34392/commits/275092cce3420fca67cbd9f58c3c7b5a9c9758e8), `DebugOptions` is an empty struct in release-mode, and the ABI is consistent in both modes because `DebugOptions` is only exposed by pointer/reference.

Reply:
> But I wonder why debug options would be ignored in release mode.

This can be discussed as part of the follow-up issue #35507